openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb

```diff
@@ -0,0 +1,343 @@
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Set up work to be executed by the aggregator and the collaborators by extending `FLSpec`:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class FederatedFlow(FLSpec):\n",
+    "    def __init__(self, model=None, optimizer=None, rounds=10, **kwargs):\n",
+    "        super().__init__(**kwargs)\n",
+    "        self.model = model\n",
+    "        self.optimizer = optimizer\n",
+    "        self.rounds = rounds\n",
+    "        self.loss = 0.\n",
+    "\n",
+    "    @aggregator\n",
+    "    def start(self):\n",
+    "        print(f'Performing initialization for model')\n",
+    "        self.collaborators = self.runtime.collaborators\n",
+    "        self.private = 10\n",
+    "        self.current_round = 0\n",
+    "        self.next(self.aggregated_model_validation, foreach='collaborators', exclude=['private'])\n",
+    "\n",
+    "    def compute_accuracy(self, data_loader):\n",
+    "        self.model.eval()\n",
+    "        test_loss = 0\n",
+    "        correct = 0\n",
+    "        with torch.no_grad():\n",
+    "            for data, target in data_loader:\n",
+    "                output = self.model(data)\n",
+    "                test_loss += F.cross_entropy(output, target, size_average=False).item()\n",
+    "                pred = output.data.max(1, keepdim=True)[1]\n",
+    "                correct += pred.eq(target.data.view_as(pred)).sum()\n",
+    "\n",
+    "        test_loss /= len(data_loader.dataset)\n",
+    "        print('\\nTest set: Avg. loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\\n'.format(\n",
+    "        test_loss, correct, len(data_loader.dataset),\n",
+    "        100. * correct / len(data_loader.dataset)))\n",
+    "        accuracy = float(correct / len(data_loader.dataset))\n",
+    "        return accuracy\n",
+    "\n",
+    "    @collaborator\n",
+    "    def aggregated_model_validation(self):\n",
+    "        print(f'Performing aggregated model validation for collaborator {self.input}, model: {id(self.model)}')\n",
+    "        self.agg_validation_score = self.compute_accuracy(self.test_loader)\n",
+    "        self.next(self.train)\n",
+    "\n",
+    "    @collaborator\n",
+    "    def train(self):\n",
+    "        # Log after processing a quarter of the samples\n",
+    "        log_threshold = .25\n",
+    "\n",
+    "        self.model.train()\n",
+    "        self.optimizer = get_optimizer(self.model)\n",
+    "        for batch_idx, (data, target) in enumerate(self.train_loader):\n",
+    "            self.optimizer.zero_grad()\n",
+    "            output = self.model(data)\n",
+    "            loss = F.cross_entropy(output, target)\n",
+    "            loss.backward()\n",
+    "            self.optimizer.step()\n",
+    "\n",
+    "            if (len(data) * batch_idx) / len(self.train_loader.dataset) >= log_threshold:\n",
```

Comment:
Instead of this complicated logic of calculating "quarter" of samples used, it is OK to print every 10 or 100 steps. We need to be cautious of all the cognitive overhead we put on the user.

Reply:
I ran experiments where there were a couple of thousand steps (small chunks), it's too spammy IMO